crates/sui/src/client_ptb/builder.rs

```diff
@@ -0,0 +1,970 @@
+                            span: *command_loc,
+                            help: None,
+                        });
+                    } else {
+                        self.errors.push(PTBError {
+                            message: format!(
+                                "Variable '{}' used again here (shadowed) for the {} time.",
+                                ident, to_ordinal_contraction(i + 1)
+                            ),
+                            span: *command_loc,
+                            help: Some("You can either rename this variable, or do not \
+                                       pass the `warn-shadows` flag to ignore these types of errors.".to_string()),
+                        });
+                    }
+                }
+            }
+        }
+
+        if !self.errors.is_empty() {
+            return Err(self.errors);
+        }
+
+        let ptb = self.ptb.finish();
+        Ok(ptb)
+    }
+
+    pub async fn build(
+        mut self,
+        program: Program,
+    ) -> Result<Tx::ProgrammableTransaction, Vec<PTBError>> {
+        for command in program.commands.into_iter() {
+            self.handle_command(command).await;
+        }
+        self.finish(program.warn_shadows_set)
+    }
+
+    /// Add a single PTB command to the PTB that we are building up.
+    /// Errors are added to the `errors` field of the PTBBuilder.
+    async fn handle_command(&mut self, sp!(span, command): Spanned<ParsedPTBCommand>) {
+        if let Err(e) = self.handle_command_(span, command).await {
+            self.errors.push(e);
+        }
+    }
+
+    // ===========================================================================
+    // Declaring and handling identifiers and variables
+    // ===========================================================================
+
+    /// Declare and identifier. This is used to support shadowing warnings.
+    fn declare_identifier(&mut self, ident: String, ident_loc: Span) {
+        let e = self.identifiers.entry(ident).or_default();
+        e.push(ident_loc);
+    }
+
+    /// Declare a possible address binding. This is used to support address resolution. If the
+    /// `possible_addr` is not an address, then this is a no-op.
+    fn declare_possible_address_binding(&mut self, ident: String, possible_addr: &Spanned<PTBArg>) {
+        match possible_addr.value {
+            PTBArg::Address(addr) => {
+                self.addresses.insert(ident.to_string(), addr.into_inner());
+            }
+            PTBArg::Identifier(ref i) => {
+                // We do a one-hop resolution here to see if we can resolve the identifier to an
+                // externally-bound address (i.e., one coming in through the initial environment).
+                // This will also handle direct aliasing of addresses throughout the ptb.
+                // Note that we don't do this recursively so no need to worry about loops/cycles.
```

Comment:
While I like this flexibility, my main concern is the need to be extra careful as one could introduce subtle bugs (e.g., I thought I re-assigned b, but forgot to, and so you get an error or unexpected result). Some examples in the docs related to this would be useful with a warning to limit the usage of this to not introduce subtle errors. 

In this case, if `--assign b @0x1` would not be there, then `--move-call b::option::is_none ...` would fail during building the PTB, correct?

Reply:
Nope! If the `--assign b @0x1` is not there, then the `--move-call b::option:...` would resolve `b` to whatever address `external` is. 

Basically, if you don't re-assign it you'll keep getting the old value, but if you re-assign it to a new value, you'll then get the new one.

Reply:
Right, I meant that it will fail if it is not a package that has module option.